src/PLCTag.jl

```diff
@@ -1,81 +1,97 @@
-			global plc_tag_get_uint8 = Cfunction{UInt8, Tuple{Int32, Cint}}(library, :plc_tag_get_uint8)
-			global plc_tag_set_uint8 = Cfunction{Cint, Tuple{Int32, Cint, UInt8}}(library, :plc_tag_set_uint8)
-			global plc_tag_get_int8 = Cfunction{Int8, Tuple{Int32, Cint}}(library, :plc_tag_get_int8)
-			global plc_tag_set_int8 = Cfunction{Cint, Tuple{Int32, Cint, Int8}}(library, :plc_tag_set_int8)
-			global plc_tag_get_float64 = Cfunction{Cdouble, Tuple{Int32, Cint}}(library, :plc_tag_get_float64)
-			global plc_tag_set_float64 = Cfunction{Cint, Tuple{Int32, Cint, Cdouble}}(library, :plc_tag_set_float64)
-			global plc_tag_get_float32 = Cfunction{Cfloat, Tuple{Int32, Cint}}(library, :plc_tag_get_float32)
-			global plc_tag_set_float32 = Cfunction{Cint, Tuple{Int32, Cint, Cfloat}}(library, :plc_tag_set_float32)
+			result = new{T}(Ref{Int32}(tag), count)
+			finalizer(t -> C.plc_tag_destroy(t[]), result.tag)
+			return result
+		end
+	end
+	
+	
+	Base.length(ref::PLCRef) = ref.count
+	Base.size(ref::PLCRef) = (ref.count,)
+	Base.eltype(ref::PLCRef{T}) where {T} = T
+	Base.iterate(ref::PLCRef, state = 1) = state > length(ref) ? nothing : (ref[state], state+1)
+	
+	function Base.getindex(ref::PLCRef, ind::Int = 1)
+		1 <= ind <= length(ref) || error("Index $(ind) is out of range for PLCRef of count $(ref.count)")
+		
+		code = C.plc_tag_read(ref.tag[], DATA_TIMEOUT)
+		code == C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(C.plc_tag_decode_error(code)))")
+		
+		val = plcget(eltype(ref), ref.tag[], ind-1)
+		return convert(eltype(ref), val)
+	end
+	
+	function Base.setindex!(ref::PLCRef, val, ind::Int = 1)
+		1 <= ind <= length(ref) || error("Index $(ind) is out of range for PLCRef of count $(ref.count)")
+		
+		code = plcset(eltype(ref), ref.tag[], ind-1, convert(eltype(ref), val))
+		code == C.PLCTAG_STATUS_OK || error("ERROR: Unable to write the data! Got error code $(code): $(unsafe_string(C.plc_tag_decode_error(code)))")
+		
+		code = C.plc_tag_write(ref.tag[], DATA_TIMEOUT)
+		code == C.PLCTAG_STATUS_OK || error("ERROR: Unable to write the data! Got error code $(code): $(unsafe_string(C.plc_tag_decode_error(code)))")
+	end
+	
+	
+	plcpath(plc::PLC, tagName::String, size::Int, count::Int) = join(map(((k, v),) -> "$(k)=$(v)", (
+		:cpu => plc.cpu,
+		:protocol => plc.protocol,
+		:gateway => plc.gateway,
+		:path => plc.path,
+		:debug => plc.debug,
+		:elem_size => size,
+		:elem_count => count,
+		:name => tagName,
+	)), '&')
+	
+	# TODO: create Bool versions as well
+	for T in (
+		:Int8, :UInt8, :Int16, :UInt16,
+		:Int32, :UInt32, :Int64, :UInt64,
+		:Float32, :Float64,
+	)
+		@eval plcget(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_get_", lowercase(String(T))))(args...; kwargs...)
+		@eval plcset(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_set_", lowercase(String(T))))(args...; kwargs...)
```

Comment:
Can you explain what this is doing? Is it defining a function `plcget` and `plcset` for each `T`?

Reply:
Correct!